### ISSUE TITLE
getcwd() fails after unsuccessful rmdir() on overlayfs

### ISSUE THREAD

**Issue by larskanis** [docker/docker#19082]
Description of problem:
---

getcwd() call fails after an unsuccessful call to rmdir() with the attempt to remove the current directory. It works however, when running on native Linux or on docker with aufs driver.

This is related to:
* https://github.com/sparklemotion/nokogiri/issues/1370
* https://github.com/docker-library/ruby/issues/55
* https://github.com/docker/docker/issues/18578

`docker version`:
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:12:04 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:12:04 UTC 2015
 OS/Arch:      linux/amd64

`docker info`:
Containers: 19
Images: 95
Server Version: 1.9.1
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.2.0-22-generic
Operating System: Ubuntu 14.04.3 LTS
CPUs: 4
Total Memory: 7.688 GiB
Name: c1255lx
ID: 322J:VBEH:MQS2:4LTK:3HMD:AHKR:TTNB:WM75:QNFO:KNCG:VPVB:Z5OY
Http Proxy: http://ccproxy:4712
WARNING: No swap limit support

`uname -a`:

Linux c1255lx 4.2.0-22-generic #27~14.04.1-Ubuntu SMP Fri Dec 18 10:57:53 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

Environment details (AWS, VirtualBox, physical, etc.):
---
physical on Intel Core i5

How reproducible:
---
100% on overlayfs

Steps to Reproduce:
---
1 Use a Dockerfile like this:
```
FROM gcc
WORKDIR /usr/src/rmdir_getcwd
COPY rmdir_getcwd.c /usr/src/rmdir_getcwd/
RUN gcc -o rmdir_getcwd rmdir_getcwd.c
CMD ["./rmdir_getcwd"]
```

2 Create `rmdir_getcwd.c` like this:
  ```C
#include <unistd.h>
#include <stdio.h>
#include <sys/stat.h>

int main(int argc, char *argv[]){
  char buf[999] = "";

  /* create a new dir */
  if( mkdir("testdir", 0777) == 0 ){
    fprintf(stderr, "\nmkdir OK\n");
  }else{
    perror("\nmkdir");
  }

  /* move into the new dir */
  if( chdir("testdir") == 0 ){
    fprintf(stderr, "\nchdir OK\n");
  }else{
    perror("\nchdir");
  }

  /* create another dir, so that testdir isn't empty */
  if( mkdir("testdir2", 0777) == 0 ){
    fprintf(stderr, "\nmkdir OK\n");
  }else{
    perror("\nmkdir");
  }

  /* first getcwd succeeds */
  if( getcwd(buf, sizeof(buf)) ){
    fprintf(stderr, "\ngetcwd: %s\n", buf);
  }else{
    perror("\ngetcwd");
  }

  fprintf(stderr, "\nrmdir(%s): ", buf);
  /* rmdir fails intentionally, with 'Directory not empty' */
  if( rmdir(buf) == 0 ){
    fprintf(stderr, "OK\n");
  }else{
    perror("rmdir");
  }

  /* second getcwd fails on overlayfs, only */
  if( getcwd(buf, sizeof(buf)) ){
    fprintf(stderr, "\ngetcwd: %s\n", buf);
  }else{
    perror("\ngetcwd");
  }

  return 0;
}
  ```

3 Run
  ```sh
docker build -t rmdir_getcwd .
docker run --rm rmdir_getcwd
  ```

Actual Results:
---
```
mkdir OK

chdir OK

mkdir OK

getcwd: /usr/src/rmdir_getcwd/testdir

rmdir(/usr/src/rmdir_getcwd/testdir): rmdir: Directory not empty

getcwd: No such file or directory
```

Expected Results:
---
```
mkdir OK

chdir OK

mkdir OK

getcwd: /usr/src/rmdir_getcwd/testdir

rmdir(/usr/src/rmdir_getcwd/testdir): rmdir: Directory not empty

getcwd: /usr/src/rmdir_getcwd/testdir
```

Additional info:
---

**Comment by tianon** [comment]
This is some awesome debugging and reproducibility work; thanks @larskanis! :+1: :heart:

**Comment by MHBauer** [comment]
This is a beautiful report.

**Comment by ncopa** [comment]
I reported it upstream: https://bugzilla.kernel.org/show_bug.cgi?id=110491

**Comment by cyphar** [comment]
I'd like to point out that `rmdir()` of the current working directory is **ALWAYS**  invalid according to POSIX. But yeah, this is clearly a kernel bug.

**Comment by ncopa** [comment]
The rmdir returns error, like it should. The problem is that getcwd fails afterwards.

**Comment by AkihiroSuda** [comment]
Fixed in Linux >= 4.5
https://github.com/torvalds/linux/commit/ce9113bbcbf45a57c082d6603b9a9f342be3ef74

**Comment by runcom** [comment]
closing as per @AkihiroSuda comment
